src/providers/token-properties-provider.ts

```diff
@@ -0,0 +1,110 @@
+import { BigNumber } from '@ethersproject/bignumber';
+import { ChainId, Token } from '@uniswap/sdk-core';
+
+import { log } from '../util';
+
+import { ICache } from './cache';
+import { ProviderConfig } from './provider';
+import { ITokenFeeFetcher, TokenFeeResult } from './token-fee-fetcher';
+import { DEFAULT_ALLOWLIST } from './token-validator-provider';
+
+const DEFAULT_TOKEN_BUY_FEE_BPS = BigNumber.from(0);
+const DEFAULT_TOKEN_SELL_FEE_BPS = BigNumber.from(0);
+
+// on detector failure, assume no fee
+const DEFAULT_TOKEN_FEE_RESULT = {
+  buyFeeBps: DEFAULT_TOKEN_BUY_FEE_BPS,
+  sellFeeBps: DEFAULT_TOKEN_SELL_FEE_BPS,
+};
+
+const DEFAULT_TOKEN_PROPERTIES_RESULT: TokenPropertiesResult = {
+  tokenFeeResult: DEFAULT_TOKEN_FEE_RESULT,
+};
+
+type Address = string;
+export type TokenPropertiesResult = {
+  tokenFeeResult: TokenFeeResult;
+};
+export type TokenPropertiesMap = Record<Address, TokenPropertiesResult>;
+
+export interface ITokenPropertiesProvider {
+  getTokensProperties(
+    tokens: Token[],
+    providerConfig?: ProviderConfig
+  ): Promise<TokenPropertiesMap>;
+}
+
+export class TokenPropertiesProvider implements ITokenPropertiesProvider {
+  private CACHE_KEY = (chainId: ChainId, address: string) =>
+    `token-properties-${chainId}-${address}`;
+
+  constructor(
+    private chainId: ChainId,
+    private tokenPropertiesCache: ICache<TokenPropertiesResult>,
+    private tokenFeeFetcher: ITokenFeeFetcher,
+    private allowList = DEFAULT_ALLOWLIST
+  ) {}
+
+  public async getTokensProperties(
+    tokens: Token[],
+    providerConfig?: ProviderConfig
+  ): Promise<TokenPropertiesMap> {
+    const tokenToResult: TokenPropertiesMap = {};
+    const addressesToFetchFeesOnchain: string[] = [];
+    const addressesRaw = this.buildAddressesRaw(tokens);
+
+    const tokenProperties = await this.tokenPropertiesCache.batchGet(addressesRaw)
+
+    // Check if we have cached token validation results for any tokens.
+    for (const address of addressesRaw) {
+      const cachedValue = tokenProperties[address];
+      if (cachedValue) {
+        tokenToResult[address] = cachedValue;
+      } else if (this.allowList.has(address)) {
+        tokenToResult[address] = DEFAULT_TOKEN_PROPERTIES_RESULT;
+      } else {
+        addressesToFetchFeesOnchain.push(address);
+      }
+    }
+
+    log.info(
+      `Got token fee results for ${
+        addressesRaw.size - addressesToFetchFeesOnchain.length
+      } tokens from cache. Getting ${
+        addressesToFetchFeesOnchain.length
+      } on-chain.`
+    );
+
+    if (addressesToFetchFeesOnchain.length > 0) {
+      const tokenFeeMap = await this.tokenFeeFetcher.fetchFees(
+        addressesToFetchFeesOnchain,
+        providerConfig
+      );
+
+      for (const address of addressesToFetchFeesOnchain) {
+        const tokenFee = tokenFeeMap[address];
```

Comment:
I was thinking to parallelize calling `TokenValidatorProvider` and `TokenPropertiesProvider` to save the latencies. But yes, then `TokenPropertiesProvider` will always go out for non-FOT tokens, and doesn't seem to worth it. Since `TokenValidatorProvider` has a in-memory cache, I think calling `TokenValidatorProvider` within `TokenPropertiesProvider` makes sense.

Reply:
@andysmith415 can you help me understand how negative cache would speed up things? I am actually unfamiliar since i've never used it. How is it different than realizing that something is empty? You just dont have to go through the array one more time to see whats left?

Reply:
I think what Andy means here is that before we fetch the token fees, we can have a cache to remember tokens that are not FOTs. Since `TokenValidatorProvider` exactly fits the negative cache purpose here (it contains the in-memory cache for FOT and non-FOT token results), it makes sense to call `TokenValidatorProvider` to check whether the token is FOT, before calling out to `TokenFeeFetcher` to fetch the token fees, even for non-FOT tokens.

Reply:
I updated the `TokenPropertiesProvider` to ensure the negative cache `TokenValidatorProvider` blocks `TokenPropertiesProvider` from querying `TokenFeeFetcher` for non-FOT tokens. Also added the unit test coverage to cover this exact negative cache case.